consensus/src/runway/backup.rs

```diff
@@ -0,0 +1,331 @@
+        self.flush()?;
+        Ok(())
+    }
+}
+
+impl<H: Hasher, D: Data, S: Signature, R: Read> UnitLoader<H, D, S> for R {
+    fn load(mut self) -> Result<Vec<UncheckedSignedUnit<H, D, S>>, LoaderError> {
+        let mut buf = Vec::new();
+        self.read_to_end(&mut buf)?;
+        let input = &mut &buf[..];
+        let mut result = vec![];
+        while !input.is_empty() {
+            result.push(<UncheckedSignedUnit<H, D, S>>::decode(input)?);
+        }
+        Ok(result)
+    }
+}
+
+fn _load_backup<H: Hasher, D: Data, S: Signature, UL: UnitLoader<H, D, S>>(
+    unit_loader: UL,
+) -> Result<(Vec<UncheckedSignedUnit<H, D, S>>, Round), LoaderError> {
+    let (rounds, units): (Vec<_>, Vec<_>) = unit_loader
+        .load()?
+        .into_iter()
+        .map(|u| (u.as_signable().coord().round(), u))
+        .unzip();
+    let next_round = if let Some(round) = rounds.into_iter().max() {
+        round + 1
+    } else {
+        0
+    };
+    Ok((units, next_round))
+}
+
+fn _on_shutdown(starting_round_tx: oneshot::Sender<Option<Round>>) {
+    if starting_round_tx.send(None).is_err() {
+        warn!(target: "AlephBFT-runway", "Coulnd not send `None` starting round.");
+    }
+}
+
+/// Loading mechanism first Unit data from `unit_loader` and awaits on response from unit collection.
+/// If loaded Units are compatible with unit collection result (meaning highest unit is from round at least
+/// unit_collection + 1) it send all loaded units by `loaded_unit_tx` and sends starting round number by
+/// `starting_round_tx`. If Units are not compatible it sends no Units and `None` by `starting_round_tx`
+pub async fn _run_loading_mechanism<H: Hasher, D: Data, S: Signature, UL: UnitLoader<H, D, S>>(
+    unit_loader: UL,
+    loaded_unit_tx: Sender<UncheckedSignedUnit<H, D, S>>,
+    starting_round_tx: oneshot::Sender<Option<Round>>,
+    highest_response_rx: oneshot::Receiver<Round>,
+) {
+    let (units, next_round) = match _load_backup(unit_loader) {
+        Ok((units, next_round)) => (units, next_round),
+        Err(e) => {
+            error!(target: "AlephBFT-runway", "unable to load unit backup: {:?}", e);
+            _on_shutdown(starting_round_tx);
+            return;
+        }
+    };
+
+    let highest_response = match highest_response_rx.await {
+        Ok(highest_response) => highest_response,
+        Err(e) => {
+            error!(target: "AlephBFT-runway", "unable to receive response from unit collections: {:?}", e);
+            _on_shutdown(starting_round_tx);
+            return;
+        }
+    };
+
+    let starting_round = next_round;
+    if starting_round < highest_response {
+        error!(target: "AlephBFT-runway", "backup and unit collection missmatch. Backup got: {:?}, collection got: {:?}", starting_round, highest_response);
+        _on_shutdown(starting_round_tx);
+        return;
+    };
+
+    for u in units {
+        if let Err(e) = loaded_unit_tx.unbounded_send(u) {
```

Comment:
I think this should happen before waiting for highest response, I don't see a reason to postpone it.

Reply:
I thought that we could send unit only if we will start the creator. But I guess sending it earlier would be better.